a11y rework of [EuiPagination] and friends

myasonik opened the pull request:
## Summary

- Closes #3244 
	- "Fixed" by moving focus away from the pagination after clicking
- Closes #2196 
	- Largely followed my initial draft structure
	- Will make a new issue for adding a header element instead of relying on `aria-*` attributes
	- May make a new issue for compressed pagination which I entirely avoided in this PR 

**EuiPagination** now sports the following changes:
- Everything is wrapped in a `<nav>` and the pages are in a `<ul>`
- Renders as `<a>` (instead of `<button>`) if there is an `id` to point to (similar idea to a skip link using `<a>`)
	- If this is the case, will move focus to the `id`
	- Also adds `aria-controls={id}`
- If the ellipses would only replace 1 page, renders the page number
- Adds more context to prev and next pagination buttons of which page they will be going to for screen readers
- Adds more context to the ellipses for screen readers
- Adjusts spacing of ellipses to match page numbers

**EuiBasicTable**, **EuiInMemoryTable** and **EuiDataGrid** pass in `aria-controls` to their pagination to take full advantage of the new structure.

## Breaking change
- **EuiPaginationButton** now requires a `pageIndex`
	- Shouldn't be a huge deal because I don't think there's a real use case to use it outside of EUI

## Checklist

- [x] Check against **all themes** for compatibility in both light and dark modes
- [x] Checked in **mobile**
- [x] Checked in **IE11** and **Firefox**
- [x] Props have proper **autodocs**
~- [ ] Added **documentation** examples~
- [x] Added or updated **jest tests**
- [x] Checked for **breaking changes** and labeled appropriately
- [x] Checked for **accessibility** including keyboard-only and screenreader modes
- [x] A [changelog](https://github.com/elastic/eui/blob/master/CONTRIBUTING.md#changelog) entry exists and is marked appropriately

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3294/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3294/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3294/

cchaos commented on src/components/pagination/_pagination_button.scss:
```diff
@@ -6,18 +6,33 @@
-  text-decoration: underline;
   font-weight: $euiFontWeightBold;
+
+  &#{&} {
```
Why this crazy selector? It just concatenates the original selector to itself like: `.euiPaginationButton-isActive.euiPaginationButton-isActive`. It seems like a hacky way to get specificity.

myasonik replied:
> It seems like a hacky way to get specificity.

That's exactly what it is 😆 First saw it many years ago on [CSS Wizardry](https://csswizardry.com/2014/07/hacks-for-dealing-with-specificity/#safely-increasing-specificity)  

Sounds like it's not a pattern in EUI so I can switch it to something else

myasonik replied:
I wasn't sure how to best fix this without hacks... Do you have a suggestion?

cchaos commented on src/components/pagination/_pagination_button.scss:
```diff
@@ -6,18 +6,33 @@
+  font-size: $euiFontSizeS;
+  padding: 0 $euiSizeS;
+  height: $euiSizeL;
+  padding-top: 6px;
```
```suggestion
  padding-top: $euiSizeM / 2;
```

myasonik commented:
## Table's focus state
I don't have anything to link to here unfortunately... I had trouble hunting down any patterns for this so this pattern is from the collective minds of an accessibility slack workspace that I'm in. All that I can link to is a spec success criterion that says everything focusable needs a visual focus state.

I see what you mean about it being visually _a lot_ though so I'm happy to try to come up with some middle ground approach:
- Maybe just reducing the outline could be enough? I could try `euiFocusRing('small')` instead of the default that I'm currently using. 
- Would take a bit more JS work but I could try not moving focus until the mouse leaves the little pagination area? 

That's all that came to mind immediately but if you have other ideas, I'm all ears.

## Compressed version
I explicitly didn't touch this to limit scope of this PR and plan to open a follow up issue. I can include it if you think it'd be weird to ship it as is.

## Number display
I agree that it's annoying that the size keeps changing as it is right now but this PR shouldn't have changed anything regarding this (other than replacing an ellipses with a number only when the ellipses replaced 1 number which shouldn't effect the overall length). 

I'd rather open up a follow up issue to resolve this than continuing to increase the scope of this PR.

myasonik commented on src/components/datagrid/data_grid.tsx:
```diff
@@ -123,25 +123,11 @@ type CommonGridProps = CommonProps &
-      }
-    | {
-        /**
-         * must provide either aria-label OR aria-labelledby as a title for the grid
-         */
-        'aria-labelledby': string;
-      });
+>;
```
I added [`OneOf`](https://github.com/elastic/eui/pull/3294/files#diff-bf9ced0406d8a848c76f950cdafdefd9R19-R20) as a new utility type to do this because the [`accessibleName`](https://github.com/elastic/eui/pull/3294/files#diff-0d500c22782b202e062244f187d2cfd0R200-R205) work below was causing **EuiDataGrid** to complain about not having `aria-label` or `aria-labelledby`.

I tried to get it working using `ExclusiveUnion`s but had no luck...

@chandlerprall Let me know what you think of this change.

chandlerprall replied:
Tested out `OneOf`, looks good! Looks like this change is more accurate/correct than the previous type, too.

snide commented:
> Do we really need a focus ring on the entire table after each page page selection? It seems excessive visually as users page through.

This is excessive, at the point of distraction, which is its own top level concern. I know I say this a lot, but there is a balance. This is a pattern I do not see regularly on even high accessibility websites.

I think ultimately only sometimes do you want to jump focus to the table itself. Sometimes you just want to continue tabbing through pagination.

chandlerprall commented on src/components/basic_table/in_memory_table.test.tsx:
```diff
@@ -924,7 +928,8 @@ describe('EuiInMemoryTable', () => {
       expect(props.onTableChange).toHaveBeenCalledTimes(0);
       component
-        .find('EuiPaginationButton[data-test-subj="pagination-button-1"]')
+        .find('[data-test-subj="pagination-button-1"]')
```
Let's change to

```suggestion
        .find('EuiButtonEmpty[data-test-subj="pagination-button-1"]')
```

and drop the `.first()`. Seeing `first` in this chain implied (to me) that having multiple buttons with that test subject is valid, when it's not true - there is only one, but the test environment finds the virtual dom node as well. Changing to explicitly look for `EuiButtonEmpty` helps narrow the test as needed but will also catch a future regression if two of these end up rendering.

chandlerprall commented on src/components/basic_table/pagination_bar.tsx:
```diff
@@ -16,6 +16,10 @@ export interface Pagination {
+  /**
+   * id of the table being controlled
+   */
+  controls?: string;
```
`controls` is a loaded term (does this show/hide something, is it passing more controls in to the bar, does it add text the user can read etc), let's rename to be more descriptive

myasonik replied:
Changed to `aria-controls`

chandlerprall commented on src/components/basic_table/pagination_bar.tsx:
```diff
@@ -28,6 +32,7 @@ export const PaginationBar = ({
   pagination,
   onPageSizeChange,
   onPageChange,
+  ...rest
```
It's only one property, prefer using the prop itself here so it doesn't make it seem like a swath of values

chandlerprall commented on src/components/common.ts:
```diff
@@ -16,6 +16,8 @@ export interface CommonProps {
 export type NoArgCallback<T> = () => T;
 
 // utility types:
+export type OneOf<T, K extends keyof T> = Omit<T, K> &
```
Add a comment showing use case & usage

chandlerprall commented:
Totally agree with Dave's assessment of that focus jumping

myasonik commented:
All the comments here seem reasonable and it's good to move forward but moving this PR back to a draft as I need to step away from it for an indefinite while to focus on some Kibana work.

myasonik commented:
Don't think I can fully get back to this PR right now but wanted to merge master into it so it doesn't fall too far behind.

I addressed just the easiest PR feedback while I had it open.

Big change (that was easy to do) is I also removed focus ring from tables. Turns out, visible focus is only required on controls (things you can interact with) which is why things like skip links don't trigger giant visible focus rings around the entire main contents of the page.

How do folks feel about moving focus to the table with the giant focus ring gone?

myasonik commented:
Don't think I can fully get back to this PR right now but wanted to merge master into it so it doesn't fall too far behind.

I addressed just the easiest PR feedback while I had it open.

Big change (that was easy to do) is I also removed focus ring from tables. Turns out, visible focus is only required on controls (things you can interact with) which is why things like skip links don't trigger giant visible focus rings around the entire main contents of the page.

How do folks feel about moving focus to the table with the giant focus ring gone?

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3294/

myasonik commented on src/components/basic_table/basic_table.tsx:
```diff
@@ -1220,11 +1225,26 @@ export class EuiBasicTable<T = any> extends Component<
+      const {
+        'aria-label': ariaLabel,
+        'aria-labelledby': ariaLabelledBy = this.captionId,
+      } = this.props;
```
Removed (labeling was a bit overeager)

myasonik commented:
Have more time now so picking this back up.

To reiterate my previous question:
> Big change (that was easy to do) is I also removed focus ring from tables. Turns out, visible focus is only required on controls (things you can interact with) which is why things like skip links don't trigger giant visible focus rings around the entire main contents of the page.

> **How do folks feel about moving focus to the table with the giant focus ring gone?**

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3294/

chandlerprall commented on src/components/pagination/pagination.tsx:
```diff
@@ -59,10 +64,48 @@ export const EuiPagination: FunctionComponent<Props> = ({
+    return button;
+  };
+  const classes = classNames('euiPagination', className);
+  const hasControl = typeof ariaControls === 'string';
```
EUI code pattern prefers eliminating the `undefined` option instead of `typeof`ing

```suggestion
  const hasControl = ariaControls !== undefined;
```

myasonik commented:
If you watch Dave's video above, you'll see the table gets reread linearly when focus is moved to it. Did some testing and research and that seems to be expected.

Also tested in VO+Safari and it seems to read the caption like 50% of the time and move focus to the wrong element the other 50% so I think we're into screen reader/browser discrepancies at this point. 

Windows screen reader testing for grid is on my to-do list so I'll review this then as well.

myasonik commented:
retest

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3294/
